Performance improvements on View Study page

zagorsky opened the pull request:
This pull request fixes a problem on `main` where if you have lots of participants in a study, the View Study page loads very slowly.  I tested it on a study with 25,000 participants: on `main`, the page was timing out with a 504 error after one minute; on this branch, the page loaded and functioned normally (pageload and all operations well under 1 second).

The big change in this PR is that instead of loading all participants in the template at pageload time, the page loads without the table, and then the table itself pulls participants as needed via AJAX.  The table uses DataTables (as it did before), but now it uses DataTables' Server-Side Processing feature.

For the most part, users shouldn't notice any significant changes, other than the page now loading much faster.  But there are a few minor changes:
* You used to be able to search/filter the table on any column, and to sort the table on any column.  Now you can only search/filter on Patient ID and Phone OS, and you can only sort the table on the four standard columns (creation date, patient ID, phone registered, and phone OS, but not on any custom field or intervention date).  This is because in the old way it was implemented, DataTables automatically handled all filtering and sorting in JavaScript, but in this new way, we need to implement every filter and sort operation as a database query in Django.  So we _can_ add that functionality for the rest of the columns, but it's not trivial.  Therefore, I want to merge the feature now, as-is, and not implement that functionality for any other columns unless and until someone asks for it.
* The "Phone registered" column previously read either "Smartphone registered" or "No smartphone registered".  Now it reads either "true" or "false".
* If a phone hasn't been registered, the "Phone OS" column previously read "_unknown_".  Now it's just blank.  But if a phone _has_ been registered, it still says "IOS" or "ANDROID", same as before.
* The View Participant page has a new URL: it uses the 8-character patient ID, instead of the integer sequential ID (the primary key from the database).  I updated all the links throughout the web app that point to the View Participant page.  But if anyone bookmarked the View Participant page for a participant, those bookmarks won't work anymore.

biblicabeebli commented:
(I don't have a populated database set up yet, I can't test the optimizations until I do, sowwy.)

I'd like to shift the `get_values_for_view_study_table` function out of the database model.  It only has a single use of `cls` and the function is only called in one place.  I've made an eli/view-study-pageload-fix branch with commits I'm referencing in this post.  The commit with hash e001f7f023fab0b891fb2cbda6efc79a892443d7 contains this change.  (I also added type annotations, so please double check if I got those wrong, I'd generally like to see those on all function declarations.)

I have a second commit, d390b70fd72afa23d1223724e8b620d13b835cd5, which is a cleanup of what looks to me like a half-finished for-loop-to-list-comprehension conversion.  Can you confirm?

I made a further performance optimization on commit 3, 733d36cefb93be58ea4de2615b0aa517c7ac352d:
There are two database queries made for every Participant object in the main loop, `field_values` and `intervention_dates`.  I've converted this to a single database query outside the loop, from which we create a lookup dictionary.  (This commit is technically experimental because I don't have the energy right now to deal with the database crap, sorry.) 

An alternative strategy would have been to use `queryset.select_related`, but I've universally found that this does not function as well as desired/predicted, and the point at which you may want to use it is also the point at which you should use the fastest django database query option: `.values_list`.

With respect to any next-step optimizations, those would require use of `queryset.values` / `queryset.values_list`, or using `queryset.only` / `queryset.defer`, to reduce the fields (columns) actually pulled from the database.  With large queries the location of hidden overhead are the following:
1) instantiating many django database objects takes some time, `.values` and `.values_list` bypasses this, `.defer` and `.only` do not.
2) My overwhelming experience has been that the bulk of unexpected overhead is inside of python converting the database's date-time representation to python `datetime` objects.  The only solution is to reduce the number of datetime objects in play.  This code makes direct use of the `created_on` attribute, but we could exclude `last_updated` on the Participant object... unless it is used later on in other code - that would make it slower.
3) Further limiting the database fields we pull in: the low-hanging fruit remaining would be using `values_list` to grab the `custom_field_val.value` and `intervention_date.date` values without pulling in the rest of the fields.  Of these InterventionDates are TimestampedModels, so that could be substantial.

Final note:
My 3rd commit happens to have factored out the existing code that was blocking an easy rewrite to _everything_ using `values_list` (the `participant.intervention_dates` and `participant.field_values` related table queries), but I don't 100% remember how to deal with the annotation on the Participant query + using `.values_list`, and I'm pretty sure you pass the output of the function to the template and have dot-accesses (`participant.attribute`) inside it, so such a rewrite would expand beyond this single function.

I'll deal with my local database issues on Monday, test my code, and do the 3rd thing in my list above... unless someone else wants to...

biblicabeebli commented:
We should profile this without the pagination to see if we can get the sortable columns back.  If the reduced database queries are a substantial save, then my guess is that of the remaining raw python compute time 80% of it is going to be just `datetime` object instantiations.  I'm counting 4 datetime objects (participant.last_updated, participant.created_on, intervention.created_on, and intervention.last_updated) of which I think we only need participant.created_on.

zagorsky commented:
@biblicabeebli
1. > I have a second commit, d390b70, which is a cleanup of what looks to me like a half-finished for-loop-to-list-comprehension conversion. Can you confirm?

    It wasn't anything half-finished, but good catch that I could eliminate a line here.  I updated it on my branch in ff22fb8.

2. > I also added type annotations, so please double check if I got those wrong, I'd generally like to see those on all function declarations.

    I added type annotations on my branch in a38c78a.  `contains_str` is a `str`, not a `bool`.

3. > I'd like to shift the get_values_for_view_study_table function out of the database model. It only has a single use of cls and the function is only called in one place.

    I see what you're saying, but I disagree because it uses a lot of `django.db.models` functions.  I think it should be left in the models file.

4. Additional performance optimizations: my attitude here is that further performance optimizations are definitely possible, but not necessary.  This PR gets the pageload down from tens of thousands of database calls down to a few dozen, and the table loads instantaneously from the user's perspective, even with 25,000 participants.  We could even further reduce the database calls in the interest of good style, but I keep getting pressured on budget.  So if you have a clean and simple improvement that you've tested, please go ahead and add it, but I think I'm out of budget to do any further optimizations.
    1. Pulling `field_values` and `intervention_dates` queries out of the main loop (your commit 733d36c): this is nifty, but I think it adds code complexity to gain performance that we don't need; I don't think it's worth the extra lines and reduction in readability.
    2. Using `select_related`- I don't think that works on `field_values` and `intervention_dates` because the keys are in the wrong direction.  `prefetch_related` would work, but I didn't have the time to figure it out.  I think using `prefetch_related` would accomplish what you're doing in 733d36c, but with simpler code.  If you know how to do that, please go ahead and show me- it'd be cool to see.
    3. Using `.values_list`: I agree, this is faster, and that's the way I was doing it before I added in querying the `field_values` and `intervention_dates`.  I abandoned using `.values_list` because I think using the database objects themselves makes the code more readable.
    4. Parsing datetimes is inefficient: I don't think this is meaningful here.  Not in comparison to making additional database calls.
    5. > We should profile this without the pagination to see if we can get the sortable columns back.

        Please, be my guest: set up your local environment and go ahead and profile it.  I believe the suggestion you're making is to abandon use of server-side processing for DataTables (the bulk of this PR), and to go back to pulling all participants on pageload, but to optimize the database queries in doing so.  It's possible that it could work!  Although then you may end up dealing with client-side performance issues; I don't know how well DataTables can handle thousands of rows.  I thought server-side processing for DataTables was the right fix for this issue, but server-side processing comes with a cost: that we need to reimplement so much of DataTables' functionality in Django queries, like column sorting.

alvinsiu commented on database/user_models.py:
```diff
@@ -167,6 +168,52 @@ def get_private_key(self):
+            # Convert the datetime object into a string in YYYY-MM-DD format
+            participant_values[0] = participant_values[0].strftime('%Y-%m-%d')
+            # Add all values for intervention dates
+            for intervention_date in participant.intervention_dates.order_by(Lower('intervention__name')):
+                if intervention_date.date is not None:
+                    participant_values.append(intervention_date.date.strftime('%Y-%m-%d'))
+                else:
+                    participant_values.append(None)
+            # Add all values for custom fields
+            for custom_field_val in participant.field_values.order_by(Lower('field__field_name')):
```
:information_source: Given your testing, not necessary to do, but running `.intervention_dates` or `order_by('interviention__name')` or `order_by('field__field_name')` will make a separate request for each participant. This can be avoided if you have more performance problems using `select_related` or `prefetch_related` (https://stackoverflow.com/a/31237071/1940450).

zagorsky replied:
Done using `prefetch_related`.  Thanks for the suggestion.  It took me a little while to figure out how to sort on a related field inside the `prefetch_related` call, but I got it done.  I checked, and it did cut the number of database calls.  This AJAX query now executes in a constant of 6 database calls, regardless of the number of participants it pulls.

alvinsiu commented on frontend/static/javascript/participants_table.js:
```diff
@@ -0,0 +1,21 @@
+    $("#participantList").DataTable({
+        "processing": true,
+        "serverSide": true,
+        "ajax": "/study/" + studyId + "/get_participants_api",
```
If you do change the API url above, don't forget to change it here too

alvinsiu commented on frontend/templates/notification_history.html:
```diff
@@ -3,11 +3,11 @@
     <p><b>Page of results:</b></p>
     <ul class="pagination" style="margin-top:0px;">
-      {% set urlbase = "/view_study/" + study.id|string + "/participant/" + participant.id|string + "/notification_history?page=" %}
+      {% set urlbase = "/view_study/" + study.id|string + "/participant/" + participant.patient_id|string + "/notification_history?page=" %}
```
:pick: The `|string` may not be necessary anymore.

zagorsky replied:
Good catch: fixed.

alvinsiu commented on frontend/templates/base.html:
```diff
@@ -4,18 +4,15 @@
-        <script src="/static/javascript/jquery-1.11.1.js" type="text/javascript"></script>
+        <script src="https://code.jquery.com/jquery-3.5.1.js" type="text/javascript"></script>
         <script src="/static/javascript/handlebars-v2.0.0.js" type="text/javascript"></script>
-        <script src="/static/javascript/jquery.dataTables.min.js" type="text/javascript"></script>
```
:lady_beetle: This is a great idea, but I think `dashboard_features.js` and `patient_dashboard_features.js` still use datatables, so you'll need to add in the js/css to those pages too.

zagorsky replied:
Whoops, good catch!  Thanks- I fixed it.

alvinsiu reviewed (approved):
A couple of small things, but otherwise looks good assuming you've done click-through testing!

zagorsky commented on api/study_api.py:
```diff
@@ -19,6 +19,32 @@ def inject_html_params():
+    sort_in_descending_order = request.args.get('order[0][dir]') == 'desc'
+    contains_string = request.args.get('search[value]')
+    total_participants_count = Participant.objects.filter(study_id=study_id).count()
+    filtered_participants_count = (Participant
+                                   .filtered_participants_for_study(study_id, contains_string)
+                                   .count())
+    data = Participant.get_values_for_view_study_table(study_id, start, length, sort_by_column_index,
+                                                       sort_in_descending_order, contains_string)
```
They're actually two different queries!  `filtered_participants_count` is the number of participants that meet the search filter.   `data` is a subset of that; it's only the ones on the current page of filtered participants.

zagorsky commented on api/study_api.py:
```diff
@@ -19,6 +19,32 @@ def inject_html_params():
+    study = Study.objects.get(pk=study_id)
+    # `draw` is passed by DataTables. It's automatically incremented, starting with 1 on the page
+    # load, and then 2 with the next call to this API endpoint, and so on.
+    draw = int(request.args.get('draw'))
+    start = int(request.args.get('start'))
+    length = int(request.args.get('length'))
```
Good point, but I actually think it's better for this to error if any of those fields is blank.  Otherwise we just get weird behavior that's hard to recognize or debug.  If `draw` doesn't get incremented, the table doesn't update.

zagorsky commented on database/user_models.py:
```diff
@@ -167,6 +168,52 @@ def get_private_key(self):
+            cls.filtered_participants_for_study(study_id, contains_string)
+            .order_by(sort_by_column)
+            .annotate(registered=ExpressionWrapper(~Q(device_id__exact=''), output_field=BooleanField()))
+            [start: start + length])
```
That's cool!  I don't think new participants are created often enough for this to matter... at least in the next year or so.  But the cursor is good to know about.  Thanks!

zagorsky commented on database/user_models.py:
```diff
@@ -167,6 +168,52 @@ def get_private_key(self):
+        basic_columns = ['created_on', 'patient_id', 'registered', 'os_type']
+        sort_by_column = basic_columns[sort_by_column_index]
+        if sort_in_descending_order:
+            sort_by_column = '-' + sort_by_column
```
Done

zagorsky commented on database/user_models.py:
```diff
@@ -167,6 +168,52 @@ def get_private_key(self):
+            sort_in_descending_order: bool,
+            contains_string: str
+    ):
+        basic_columns = ['created_on', 'patient_id', 'registered', 'os_type']
+        sort_by_column = basic_columns[sort_by_column_index]
```
Thanks! That's a good idea, but I'm already using the list of `basic_columns` in this function.  I'm just going to leave it as-is, but keep it in mind next time I design something like this.

zagorsky commented on database/user_models.py:
```diff
@@ -167,6 +168,52 @@ def get_private_key(self):
+        query = (
+            cls.filtered_participants_for_study(study_id, contains_string)
+            .order_by(sort_by_column)
+            .annotate(registered=ExpressionWrapper(~Q(device_id__exact=''), output_field=BooleanField()))
```
Thanks!  Done.

zagorsky commented on database/user_models.py:
```diff
@@ -167,6 +168,52 @@ def get_private_key(self):
+            sort_by_column_index: int,
+            sort_in_descending_order: bool,
+            contains_string: str
+    ):
```
Great point.  I made this change, and it's a lot cleaner now.  Thank you!

zagorsky commented on api/study_api.py:
```diff
@@ -19,6 +19,32 @@ def inject_html_params():
     }
 
 
+@study_api.route('/study/<string:study_id>/get_participants_api', methods=['GET'])
```
I agree, this is a good suggestion!  But I don't want to start now.  We have lots of endpoints where API version management is more critical: specifically, the API endpoints for the mobile apps.  I think we should do this in the future, though.

zagorsky commented:
Responded to all review comments from @alvinsiu and @biblicabeebli, tested locally, and tested most of the code (although not the final version) on the deployment with 25,000 participants in one study.  Merging to `main` now and deploying.
